typing/ctype.ml

```diff
@@ -2915,30 +2915,41 @@ let is_contractive env p =
 
 exception Occur
 
-let rec occur_rec env allow_recursive visited ty0 ty =
-  if eq_type ty ty0 then raise Occur;
-  match get_desc ty with
-    Tconstr(p, _tl, _abbrev) ->
-      if allow_recursive && is_contractive env p then () else
-      begin try
-        if TypeSet.mem ty visited then raise Occur;
-        let visited = TypeSet.add ty visited in
-        iter_type_expr (occur_rec env allow_recursive visited ty0) ty
-      with Occur -> try
-        let ty' = try_expand_head try_expand_safe env ty in
-        (* This call used to be inlined, but there seems no reason for it.
-           Message was referring to change in rev. 1.58 of the CVS repo. *)
-        occur_rec env allow_recursive visited ty0 ty'
-      with Cannot_expand ->
-        raise Occur
-      end
-  | Tobject _ | Tvariant _ ->
-      ()
-  | _ ->
-      if allow_recursive ||  TypeSet.mem ty visited then () else begin
-        let visited = TypeSet.add ty visited in
-        iter_type_expr (occur_rec env allow_recursive visited ty0) ty
-      end
+let occur_rec env allow_recursive visited ty0 ty =
+  let rec check_for_occurrences_in_ty ~visited ty =
+    let check_in_children ~visited ty =
+      fold_type_expr
+        (fun visited ty' -> check_for_occurrences_in_ty ~visited ty')
+        visited
+        ty
+    in
+    if eq_type ty ty0 then raise Occur;
+    if TypeSet.mem ty visited then visited
```

Comment:
This looks like a change in behavior relative to the old version: the old one would raise Occur in the Tconstr case.

Reply:
If you're referring to line 2924 of the old code, that raise is just a weird form of control flow. The `Occur` is immediately caught on line 2927.

Reply:
Right... but the old code would then expand and try again. The new version just silently succeeds. So I still think there's a difference here.

Reply:
Ah okay I see your qualm with this. I've pushed a change that is more conservative but still fixes the performance problem that was of concern.